homeassistant/components/sungrow/config_flow.py

```diff
@@ -0,0 +1,83 @@
+"""Config flow for Sungrow Solar Energy integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from SungrowModbusClient import Client, sungrow_register
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required("host"): str,
+        vol.Optional("port"): int,
+    }
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    port = data.get("port", 502)
+    client = Client(sungrow_register.regmap, data["host"], port)
+
+    client.load_register()
+    if "4990 ~ 4999 - Serial number" not in client.inverter:
+        raise InvalidAuth
+
+    # Return info that you want to store in the config entry.
+    return {
+        "title": "Sungrow " + client.inverter["4990 ~ 4999 - Serial number"].decode()
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Sungrow Solar Energy."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
+        errors = {}
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except CannotConnect:
+            errors["base"] = "cannot_connect"
+        except InvalidAuth:
+            errors["base"] = "invalid_auth"
+        except Exception:  # pylint: disable=broad-except
```

Comment:
Probably this is not necessary, as there are only a few errors you need to catch.